README.rst
Frecency for Python and Django
==================================

.. image:: https://travis-ci.org/mokelly/frecency.svg?branch=master
    :target: https://travis-ci.org/mokelly/frecency

Efficient data structures for recency-weighted counting and measurements.






HISTORY.rst
.. :changelog:

****************
History
****************

0.1.0 (2014-10-20)
=====================

* First release on GitHub



test/test_frecency.py
from __future__ import division

import random
import time
from collections import Counter


import numpy

from frecency import *
from frecency.weighted_average import WeightedAverage
from frecency.bootstrap import Bootstrap


def approx_equal(float1, float2, tol=0.001):
    result = (1. - tol) * float2 <= float1 <= (1. + tol) * float2
    return result


def test_frecency_weighting():
    now = time.time()
    timescale = 10.  # 10 seconds
    soon = now + timescale
    later = soon + timescale
    f1 = Frecency(timescale=timescale)
    f2 = Frecency(timescale=timescale)
    f1.increment(event_time=now)
    f2.increment(event_time=soon)
    value1 = f1.get_present_weight(event_time=now)
    value2 = f2.get_present_weight(event_time=soon)
    assert approx_equal(value1, 1.)
    assert approx_equal(value2, 1.)
    # Values at each event time should be equal
    assert approx_equal(value1, value2)
    # At time "soon", value2 should have twice the weight
    value3 = f1.get_present_weight(event_time=soon)
    assert approx_equal(value3 * 2., value2)
    # At any later time, the ratio should remain constant
    value4 = f1.get_present_weight(event_time=later)
    value5 = f2.get_present_weight(event_time=later)
    assert approx_equal(value4 * 2., value5)
    # Weights should decay exponentially
    assert approx_equal(value4, 0.25)
    assert approx_equal(value5, 0.5)
    # Present values should accumulate linearly
    f1.increment(9)
    value6 = f1.get_present_weight(event_time=now)
    assert approx_equal(value6, 10.)


def test_timescales():
    now = time.time()
    timescale = 10.  # 10 seconds
    soon = now + timescale
    later = soon + timescale
    f1 = Frecency(timescale=timescale)
    f2 = Frecency(timescale=timescale / 2.)  # Decays twice as fast
    f1.increment(event_time=now)
    f2.increment(event_time=now)
    value1 = f1.get_present_weight(event_time=soon)
    value2 = f2.get_present_weight(event_time=soon)
    assert approx_equal(value1, value2 * 2.)


def test_weighted_average():
    now = time.time()
    timescale = 10.  # 10 seconds
    soon = now + timescale
    w = WeightedAverage(timescale=timescale)
    num_samples = 100000
    samples = []
    for i in range(num_samples):
        sample = random.random()  # Mean 0.5, std 12**-0.5
        w.add_sample(sample, event_time=now)
        samples.append(sample)
    mean, std, uncertainty = w.get_mean_std_uncertainty(event_time=now)
    assert approx_equal(mean, numpy.mean(samples))
    assert approx_equal(std, numpy.std(samples))
    assert approx_equal(uncertainty, std / num_samples ** 0.5)
    # After some time, uncertainty has increased, but mean and std are stable
    mean2, std2, uncertainty2 = w.get_mean_std_uncertainty(event_time=soon)
    assert approx_equal(mean, mean2)
    assert approx_equal(std, std2)
    assert approx_equal(uncertainty * 2 ** .5, uncertainty2)


def test_bootstrap():
    timescale = 1.  # 1 second
    num_samples = 2 ** 20
    tol = 0.05
    b = Bootstrap(timescale=timescale)
    start_time = time.time()
    b.add_sample(1., event_time=start_time)
    b.add_sample(2., event_time=start_time + timescale, weight=2.)
    b.add_sample(3., event_time=start_time + 2 * timescale)
    b.add_sample(4., event_time=start_time + 3 * timescale)
    samples = b.get_samples(num_samples)
    sample_counts = Counter(samples)
    assert len(samples) == num_samples
    # Check that the ratios of the sample counts are as expected
    assert sample_counts[4] > 0
    assert approx_equal(sample_counts[4], sample_counts[3] * 2, tol=tol)
    assert approx_equal(sample_counts[4], sample_counts[2] * 2, tol=tol)
    assert approx_equal(sample_counts[4], sample_counts[1] * 8, tol=tol)

    # Test resampling with a predicate
    b2 = b.resample(lambda s: s>2)
    samples2 = b2.get_samples(num_samples)
    sample_counts2 = Counter(samples2)
    assert sample_counts2[4] > 0
    assert approx_equal(sample_counts2[4], sample_counts2[3] * 2, tol=tol)
    assert sample_counts2[2] == 0
    assert sample_counts2[1] == 0
    



frecency/frecency.py
"""
Frecency library.
==================================

Implementation of exponentially weighted frecency (similar to http://mathb.in/708 )


* TODO: Fast containers (Numpy array?  heapq?) with sortability, random sampling, histogram, percentiles, top-k
* TODO: https://read-the-docs.readthedocs.org/en/latest/getting_started.html
* TODO: Automatic half-life determination based on overall usage.
* TODO: Standard arithmetic operations, where 2nd argument can be Frecency or number
* TODO: Overflow and underflow detection
* TODO: Allow timescale and time0 to be changed on the fly (preserving present value)


by Michael J.T. O'Kelly, 2013-05-05
"""
from __future__ import division
from __future__ import print_function
from past.builtins import cmp
from past.utils import old_div

import time
import warnings

from numpy import logaddexp2, log2
import numpy


DEFAULT_TIME0 = time.mktime((2017, 1, 1, 0, 0, 0, 0, 0, 0))  # Arbitrarily chosen base time for exponential weight normalization
DEFAULT_TIMESCALE = 24. * 60. * 60.


class Frecency(object):
    """Exponentially weighted frecency measure"""
    def __init__(self,
                 timescale=DEFAULT_TIMESCALE,
                 start_value=0.,
                 time0=DEFAULT_TIME0,
                 suppress_warnings=False,
                 fast_comparisons=True):
        """
        * *timescale* is the halflife of events, in seconds.  With the default (24 hours)
          an event now counts twice as much as an event 24 hours ago.
        * *start_value* sets the starting weighted frecency at the present time.  (Defaults to 0.)
        * *time0* is the base time (since the epoch) for exponential weight normalization.  *Not generally worthwhile to change from default*
        * *suppress_warnings* turns off any possible warnings from misuse of this object. (Defaults to False)
        * *fast_comparisons*  If True, performs fastest possible comparison in calls to __cmp__ without sanity checking.  Overrides *suppress_warnings*
        """
        self.timescale = timescale
        self.time0 = time0
        self.fast_comparisons = fast_comparisons
        if fast_comparisons:
            self.suppress_warnings = True
        else:
            self.suppress_warnings = suppress_warnings
        self.log2_value = -numpy.Infinity  # Frecency value is stored in log2 scale
        if start_value:
            self.increment(start_value)

    def increment(self, value_added=1., event_time=None):
        """
        Increment frecency, with value_added weighted according to time of observation.
        
        * *value_added* is the number or weight of current events to add to the Frecency counter.  (e.g., 1 for one view)
        * *event_time* can be used to set the time at which the new value was added; otherwise, the present time is used.
        """
        if not event_time:
            event_time = time.time()
        log2_weight_added = (event_time - self.time0) / self.timescale + log2(value_added)
        self.log2_value = logaddexp2(self.log2_value, log2_weight_added)  # All calculations in log2 space to avoid overflow

    def get_present_weight(self, event_time=None):
        """Return the equivalent number of instantaneous events to get the current frecency, at event_time (if given) or present time."""
        if not event_time:
            event_time = time.time()
        present_weight = 2. ** (self.log2_value - (event_time - self.time0) / self.timescale)
        return present_weight

    def __cmp__(self, frec2):
        """
        Compare the present weighted values of two Frecency objects.
        **Note:** Comparison is of doubtful value between Frecencies with different timescales.  A
        Warning will trigger if this is done, unless suppress_warnings is True.
        If fast_comparisons == True, timescale and time0 are assumed to be equal
        without checking.
        If fast_comparisons == False, a completely general comparison
        of present weights is used.
        """
        if self.fast_comparisons:
            return cmp(self.log2_value, frec2.log2_value)
        else:
            if self.timescale != frec2.timescale:
                if not self.suppress_warnings:
                    warnings.warn("Different frecency timescales, {} vs. {}.  Comparison may be meaningless.".format(self.timescale, frec2.timescale))
            present_weight1 = self.get_present_weight()
            present_weight2 = frec2.get_present_weight()
            return cmp(present_weight1, present_weight2)


if __name__ == '__main__':
    f1 = Frecency()
    print(f1.get_present_weight(), f1.log2_value)
    f1.increment()
    print(f1.get_present_weight(), f1.log2_value)
    f1.increment(3)
    print(f1.get_present_weight(), f1.log2_value)

    yesterday = time.time() - 24 * 60 * 60
    f1 = Frecency()
    print(f1.get_present_weight(), f1.log2_value)
    f1.increment(event_time=yesterday)
    print(f1.get_present_weight(), f1.log2_value)
    f1.increment()
    print(f1.get_present_weight(), f1.log2_value)
    for i in range(10):
        f1.increment()
        print(f1.log2_value, end=' ')
    print()

    the_future = time.time() + 10 * 365 * 24 * 60 * 60
    f1 = Frecency(timescale=1.)
    print(f1.get_present_weight(), f1.log2_value)
    f1.increment(event_time=the_future)
    print(f1.get_present_weight(event_time=the_future), f1.log2_value)

    f1 = Frecency(suppress_warnings=False)
    f2 = Frecency()
    f2 = Frecency(timescale=1.,
                  fast_comparisons=False,
                  suppress_warnings=False)
    f1.increment(10, yesterday)
    while f1 > f2:
        f2.increment(1.)
        print(f1.get_present_weight(), f2.get_present_weight())








frecency/bootstrap.py
"""
Data structure for performing sampling of a boostrap random variable
with exponentially weighted samples in time.


TODO: Implement purge of low-weight samples (automatic or manual) to reduce memory usage


by Michael J.T. O'Kelly, 2014-04-09
"""
from __future__ import print_function
from __future__ import absolute_import

import random
import bisect

import numpy
import numpy.random

from . import frecency


class Bootstrap(object):
    """Bootstrap random variable with exponentially weighted
    samples in time.
    This code is optimized for samples which are continuous variables.  For large
    numbers of low-cardinality samples, it could be greatly sped up.
    """
    def __init__(self, timescale=frecency.DEFAULT_TIMESCALE):
        """* *timescale* is the halflife of events, in seconds.  With the default (24 hours)
        an event now counts twice as much as an event 24 hours ago.
        """
        self.timescale = timescale
        self.sample_list = []
        self.weight_cummulant_list = []
        self.weight_list = []
        self.event_time_list = []
        self.total_weight = frecency.Frecency(timescale)

    def add_sample(self, sample, weight=1.0, event_time=None):
        """Add an observed sample (which may be any object) to the bootstrap.
        * *weight* is the relative weight assigned to this sample.
        * *event_time* is the epoch time of this sample (if different from 'now')
        """
        self.sample_list.append(sample)
        self.total_weight.increment(weight, event_time=event_time)
        # Get the internal current value representing the total weight
        weight_cummulant = self.total_weight.log2_value
        self.weight_cummulant_list.append(weight_cummulant)
        # Record weight and event_time so we can resample if desired
        self.weight_list.append(weight)
        self.event_time_list.append(event_time)

    def get_sample(self):
        """This is faster than get_samples() for n==1"""
        # A random number in [0,1] represents what fraction of the cummulant we're seeking for our sample
        seek_fraction = random.random()
        # Since the cummulants are all in log2-space, we need to convert.
        # seek_position will be a float in (-Infinity, total_weight)
        seek_position = self.total_weight.log2_value + frecency.log2(seek_fraction)
        seek_index = bisect.bisect(self.weight_cummulant_list, seek_position)
        sample = self.sample_list[seek_index]
        return sample

    def get_samples(self, num_samples):
        """Efficient sampling for num_samples >> 1"""
        # We perform num_samples samples simultaneously
        # A random number in [0,1] represents what fraction of the cummulant we're seeking for our sample
        seek_fraction_array = numpy.random.rand(num_samples)
        # Since the cummulants are all in log2-space, we need to convert.
        # seek_position will be a float in (-Infinity, total_weight)
        seek_position_array = self.total_weight.log2_value + numpy.log2(seek_fraction_array)
        seek_indexes = numpy.searchsorted(self.weight_cummulant_list, seek_position_array)
        samples = [ self.sample_list[seek_index] for seek_index in seek_indexes ]
        return samples

    def resample(self, filter_func=None, timescale=None):
        """Build a new Bootstrap object containing only samples for which
        filter_func(sample) is True (if given) and with new timescale (if given)."""
        if timescale is None:
            timescale = self.timescale
        new_bootstrap = Bootstrap(timescale)
        for i in range(len(self.sample_list)):
            sample = self.sample_list[i]
            # Include this sample if there's no filter_func, or if it's True on this sample
            if not filter_func or filter_func(sample):
                weight = self.weight_list[i]
                event_time = self.event_time_list[i]
                new_bootstrap.add_sample(sample, weight=weight, event_time=event_time)
        return new_bootstrap


if __name__=='__main__':
    import time
    from collections import Counter
    b = Bootstrap(timescale=1.)
    start_time = time.time()
    b.add_sample(1., event_time=start_time)
    b.add_sample(2., event_time=start_time + 1, weight=2.)
    b.add_sample(3., event_time=start_time + 2)
    b.add_sample(4., event_time=start_time + 3)
    samples = b.get_samples(2 ** 20)
    sample_counts = Counter(samples)
    print("time passed:", time.time() - start_time)
    print(sample_counts)

    b2 = b.resample(lambda s: s>2)
    samples2 = b2.get_samples(2 ** 20)
    sample_counts2 = Counter(samples2)
    print(sample_counts2)


frecency/django/field.py
"""
Module for a Django Field subclass for storing Frecency objects
in your database (with Float at the internal representation).

(https://docs.djangoproject.com/en/dev/howto/custom-model-fields/)

TODO: Add tests along these lines:
https://github.com/gintas/django-picklefield/blob/master/src/picklefield/tests.py


by Michael J.T. O'Kelly, 2014-04-09
"""

from django.db import models
import numpy

import frecency
from future.utils import with_metaclass


DEFAULT_TIMESCALE = 24. * 60. * 60.


class FrecencyField(with_metaclass(models.SubfieldBase, models.FloatField)):

    description = "Field for storing Frecency objects"

    def __init__(self, timescale=DEFAULT_TIMESCALE, *args, **kwargs):
            self.timescale = timescale
            # Allow None values by default, because we will interpret 
            # them as -Infinity (which is not supported by FloatField).
            kwargs.setdefault('null', True)
            kwargs.setdefault('blank', True)
            kwargs.setdefault('editable', False)
            super(FrecencyField, self).__init__(*args, **kwargs)

    def deconstruct(self):
            name, path, args, kwargs = super(FrecencyField, self).deconstruct()
            # Only include kwarg if it's not the default
            if self.timescale != DEFAULT_TIMESCALE:
                kwargs['timescale'] = self.timescale
            return name, path, args, kwargs

    def to_python(self, value):
        if isinstance(value, frecency.Frecency):
            return value
        elif isinstance(value, str):
            value = float(value)
        elif value is None:
            value = -numpy.Infinity

        # 'value' must be a float at this point
        f = frecency.Frecency(timescale=self.timescale)
        f.log2_value = value  # Overwrite the Frecency internal with the value from the database
        return f

    def get_prep_value(self, value):
        internal_value = value.log2_value
        return internal_value






frecency/weighted_average.py
"""
Data structures for taking exponentially weighted averages of sampled variables
over time.

TODO: Graphical demo showing noisy data with abrupt shift in mean.
    Demonstrate that a longer timescale takes longer to adjust, but is
    smoother when there are no discontinuities.


by Michael J.T. O'Kelly, 2014-04-09
"""
from __future__ import division
from __future__ import print_function
from __future__ import absolute_import

from . import frecency


OFFSET_DEFAULT = -1e-5  # Constant to help handle negative samples


class WeightedAverage(object):
    """Exponentially weighted average of a variable sampled over time.
    NOTE: The underlying data structure requires that all samples be greater
    than the initially given 'offset' value."""
    # TODO: Arbitrary values could potentially be accepted by keeping
    # an offset value
    def __init__(self, timescale=frecency.DEFAULT_TIMESCALE, offset=OFFSET_DEFAULT):
        """* *timescale* is the halflife of events, in seconds.  With the default (24 hours)
            an event now counts twice as much as an event 24 hours ago.
        * *offset* is a value lower (by at least a little) than any sample
            to be observed."""
        self.n_sum = frecency.Frecency(timescale)  # Weight accumulator
        self.x_sum = frecency.Frecency(timescale)  # Variable accumulator
        self.x2_sum = frecency.Frecency(timescale)  # Variable squared accumulator
        self.timescale = timescale
        self.offset = offset

    def _apply_sample_offset(self, sample):
        """Applies the current offset value to the sample, adjusting it
        if necessary, and returns offset_sample.

        Since Frecency() objects can only handle positive values, we
        maintain an offset value that allows us to correct
        for negative samples in get_mean_std_uncertainty()."""
        if sample <= self.offset:
            raise ValueError("Samples must be greater than the initialized offset value", sample, self.offset)
        offset_sample = sample - self.offset
        return offset_sample

    def add_sample(self, sample, weight=1.0, event_time=None):
        """Incorporate a new sample into the weighted average.

        * *sample* is one sampled value of the variable.  (It must be
            greater than 'offset'.)
        * *weight* is the relative weight assigned to this sample.
        * *event_time* is the epoch time of this sample (if different from 'now')
        """
        offset_sample = self._apply_sample_offset(sample)
        self.n_sum.increment(weight, event_time=event_time)
        self.x_sum.increment(offset_sample, event_time=event_time)
        self.x2_sum.increment(offset_sample ** 2, event_time=event_time)

    def get_mean_std_uncertainty(self, event_time=None):
        """Returns (mean, std, uncertainty) tuple of present best estimates.

        *uncertainty* is the estimated error of the *mean*.

        The accuracy of these estimates becomes dubious when
        the sample rate is smaller than the timescale."""
        present_n = self.n_sum.get_present_weight(event_time=event_time)
        present_x = self.x_sum.get_present_weight(event_time=event_time)
        present_x2 = self.x2_sum.get_present_weight(event_time=event_time)

        offset_mean = present_x / present_n
        # Calculate the variance from the X^2 cumulant
        # (http://en.wikipedia.org/wiki/Variance#Definition)
        variance = present_x2 / present_n - offset_mean ** 2
        # http://en.wikipedia.org/wiki/Variance#Sample_variance
        # TODO: Sample bias correction.  This was attempted:
        # variance = present_n / (present_n - 1.)
        # but that didn't work in the exponential weighting framework
        std = variance ** 0.5
        uncertainty = std / present_n ** 0.5
        mean = offset_mean + self.offset  # Unapply the offset correction

        return (mean, std, uncertainty)


if __name__=='__main__':
    import random
    import time
    w = WeightedAverage(timescale=0.01, offset=-10.)
    start_time = time.time()
    for i in range(100000):
        # sample = random.gauss(10., 1.)
        sample = random.gauss(0., 1.)
        w.add_sample(sample)
    print("time passed:", time.time() - start_time)
    print(w.get_mean_std_uncertainty())
